### PR TITLE
#3 Adding 'Pre' Hooks for User Events

### PR DESCRIPTION
**Pull request by AndyM84**
Adds the following 'PRE' hooks to `UserEvents`:
```php
class UserEventTypes extends EnumBase {
  const int CONFIRM_PRE       = 9;
  const int CREATE_PRE        = 10;
  const int DELETE_PRE        = 11;
  const int LOGIN_PRE         = 12;
  const int LOGOUT_PRE        = 13;
  const int REGISTER_PRE      = 14;
  const int RESETPASSWORD_PRE = 15;
  const int UPDATE_PRE        = 16;
}
```

Each of these gets executed before any change in their respective user event.  Each of them has a consumable dispatch, which when consumed will tell the event system that the event should NOT proceed.

Still need to make documentation for this (and the rest of the event system), but I'm going to go ahead and move forward with this to keep things moving.

Changes include:
* Created 'pre' event dispatch classes
* Moved all event dispatch classes into separate class file (`~/inc/classes/UserEventDispatches.cls.php`)
* Added new `UserEventTypes` consts for 'pre' events
* Added `UserEventTypes::touchPreEvent()` method to help w/ calling 'pre' events
* Implemented calls to `UserEventTypes::touchPreEvent()` for all user events

### CODE REVIEW

File: inc/utilities/UserEvents.utl.php [hunk 1]
```diff
@@ -716,7 +447,6 @@ public function doDelete(ParameterHelper $params) : ReturnHelper {
 			(new UserRoles($this->db, $this->log))->deleteAllForUser($user->id);
 			(new UserRelations($this->db, $this->log))->deleteAllForUser($user->id);
 			(new UserProfiles($this->db, $this->log))->deleteAllForUser($user->id);
-			(new UserDevices($this->db, $this->log))->deleteAllForUser($user->id);
```
Comment (mhykes) [inline inc/utilities/UserEvents.utl.php:719]: Why remove the removal of devices?  This change doesn't seem relevant to the addition of the _PRE events.

**Review by mhykes — APPROVED** [review]
Overall this looks.  It adds a more nuanced way to prevent user events from happening.  The EVENT_PRE naming scheme feels a bit odd, but I can't come up with a better one.

File: inc/utilities/UserEvents.utl.php (hunk 1 above)
Comment (AndyM84) [reply inc/utilities/UserEvents.utl.php:719]: I just found this in here as a leftover of another effort I'd done to remove the Devices stuff.

**Comment by mhykes** [comment]
Ok.  Having confirmed that this change was intentional, please proceed to
merge this PR.
